Library/Homebrew/utils/lock.sh

```diff
@@ -0,0 +1,40 @@
+# Ref: http://wiki.bash-hackers.org/howto/mutex
+
+HOMEBREW_LOCK_DIR="$HOMEBREW_LIBRARY/lock.d"
+HOMWBREW_LOCK_PID_FILE="$HOMEBREW_LOCK_DIR/PID"
+
+lock() {
```

Comment:
I'm in favor of the current implementation. Using an atomic `mkdir` as a mutex is a pretty standard technique in shell scripting as far as I know.